pkg/manager/config/manager.go

```diff
@@ -59,21 +71,61 @@ func (srv *ConfigManager) Init(addrs []string, cfg config.ConfigManager, logger
 
 	etcdClient, err := clientv3.New(etcdConfig)
 	if err != nil {
-		return errors.WithMessage(err, "create etcd config center error")
+		return errors.Wrapf(err, "create etcd config center error")
 	}
 	srv.etcdClient = etcdClient
 	srv.kv = clientv3.NewKV(srv.etcdClient)
 
+	ctx, cancel := context.WithCancel(ctx)
+	srv.cancel = cancel
+
+	srv.initProxyConfig(ctx)
+
 	return nil
 }
 
+func (e *ConfigManager) watch(ctx context.Context, ns, key string, f func(*zap.Logger, *clientv3.Event)) {
+	wkey := path.Join(e.basePath, ns, key)
+	logger := e.logger.With(zap.String("comp", wkey))
+	e.wg.Run(func() {
+		wch := e.etcdClient.Watch(ctx, wkey)
+		ticker := time.Tick(DefaultWatchLease)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker:
+				resp, err := e.kv.Get(ctx, path.Join(e.basePath, ns, wkey))
+				if err != nil {
+					logger.Warn("failed to poll", zap.Error(err))
+					break
+				}
+				if len(resp.Kvs) != 1 {
+					logger.Warn("failed to poll", zap.Error(ErrNoOrMultiResults))
+					break
+				}
+			case res := <-wch:
+				if res.Canceled {
```

Comment:
When will it be canceled?

Reply:
When there is an error on etcd.

Reply:
> is it possible to be in a dead loop if etcd always fails

This is expected. It should be in the loop, until context is canceled. But yes, it should not consume much CPU usage.

The only case should be network rejection, where server rejected connections from the client actively. Common firewall will only drop packets. So it is a very rare case.

But anyway, sleeping for seconds or 1 min does not hurt.